compiler/rustc_middle/src/mir/interpret/allocation.rs

```diff
@@ -385,12 +385,9 @@ impl<Tag: Provenance, Extra> Allocation<Tag, Extra> {
     ) -> AllocResult {
         assert!(self.mutability == Mutability::Mut);
 
-        let val = match val {
-            ScalarMaybeUninit::Scalar(scalar) => scalar,
-            ScalarMaybeUninit::Uninit => {
```

Comment:
Yeah it requires knowledge that the only two enum variants are `Scalar` and `Uninit`. `let else` hides all non matching patterns into a single `else` clause. This also has maintainability issues as you won't get an non exhaustive match error if you add a variant that might be relevant for the code. I have added variants in the past and the non exhaustive match error was very helpful because it pointed me to all those places, some you might never have thought of, that needed change. Generally this second issue exists with `if let` as well though.

Anyways, I'll change it back.

Reply:
Not related to this commit, but I wonder if this type would be more readable as a newtype for `Option`: `struct ScalarMaybeUninit(Option<...>)`.

Reply:
I think having a variant named `Uninit` is quite useful and evocative. Code that `match`es on this needs to be very careful to correctly implement the subtle semantics of uninit values, so I'd rather not mix that up with an innocent-looking `None`.